Chained windows (#29)
ververica/flink-sql-cookbook: recipes/22/22_chained_windows.md

# 22 Chained (Event) Time Windows

:bulb: This example will show how to efficiently aggregate time series data on two different levels of granularity.

> The source table (`server_logs`) is backed by the [`faker` connector](https://github.com/knaufk/flink-faker), which continuously generates rows in memory based on Java Faker expressions.
> 
> Let's say based on our `server_logs` table we would like to compute the average request size over one minute **as well as five minute (event) windows.** 
> For this you could run two queries, similar to the one in [Aggregating Time Series Data](../06/06_group_by_window.md). 
> The script and the resulting job graph of this approach can be found at the end of this page.
> 
> In the main part we will follow a slightly more efficient approach, that chains the two aggregation: the output of the one minute aggregation serves as the input for the five minute aggregation.
> 
> We then use a [Statements Set](../17/17_statement_sets.md) to write out the two result tables. 
> To keep this example self-contained we use two tables of type `blackhole` instead of `kafka`, `filesystem` or any of the other [connectors](https://ci.apache.org/projects/flink/flink-docs-release-1.12/dev/table/connectors/). 

Reviewer: ```suggestion
The source table (`server_logs`) is backed by the [`faker` connector](https://github.com/knaufk/flink-faker), which continuously generates rows in memory based on Java Faker expressions.

Based on our `server_logs` table we would like to compute the average request size over one minute **as well as five minute (event) windows.** 
For this, you could run two queries, similar to the one in [Aggregating Time Series Data](../06/06_group_by_window.md). 
At the end of the page is the script and resulting JobGraph from this approach. 

In the main part, we will follow a slightly more efficient approach that chains the two aggregations: the one-minute aggregation output serves as the five-minute aggregation input.

We then use a [Statements Set](../17/17_statement_sets.md) to write out the two result tables. 
To keep this example self-contained, we use two tables of type `blackhole` instead of `kafka`, `filesystem`, or any other [connectors](https://ci.apache.org/projects/flink/flink-docs-release-1.12/dev/table/connectors/). 
```
